In [2]:
import lib as lib
from inputs import *
import numpy as np

# Timing of the parallelization advantage

We the performance of the parallelization with different number of cores, random walkers, and steps for each random walker (using the Harmonic Oscillator system). 

In [4]:
# Input parameters
E_local_f = E_local_Harmonic_Oscillator
prob_density = prob_density_Harmonic_Oscillator
dim = 1 # dimension of configuration space
alpha = np.array([0.65])
trial_move = 1.2222

N_skip = 1000
L_start = 5

## Analysis of the N_walkers

Using N_steps = 25000, the results are

| N_walkers | time [s] (1 core) | time [s] (4 cores) |
| --- | --- | --- |
| 250 | 1.75 | 2.07 |
| 1000 | 3.76 | 3.02 |
| 4000 | 9.75 | 6.38 |

from which we can see that if the number of walkers is high, there is an increase of the performance using parallelization (although it is not by a factor of 4).

In [20]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

1.75 s ± 71.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [6]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

2.07 s ± 71 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [7]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*4
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

3.76 s ± 176 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [8]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*4
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

3.02 s ± 111 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [9]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*16
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

9.75 s ± 224 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [10]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*16
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

6.38 s ± 45.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Analysis of the N_steps

Using N_walkers = 1000, the results are

| N_steps | time [s] (1 core) | time [s] (4 cores) |
| --- | --- | --- |
| 2500 | 0.43 | 0.51 |
| 25000 | 3.59 | 3.16 |
| 50000 | 6.58 | 5.74 |
| 75000 | 9.80 | 8.61 |

from which we can see that the decrease in run time using more cores is not as good as when varying the number of walkers, but still there is an advantage compared to using a single core. 

In [11]:
# Monte Carlo integration params
N_steps = 2500
N_walkers = 250*4
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

430 ms ± 33 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [12]:
# Monte Carlo integration params
N_steps = 2500
N_walkers = 250*4
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

512 ms ± 69.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [13]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*4
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

3.59 s ± 182 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [14]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*4
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

3.16 s ± 267 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [15]:
# Monte Carlo integration params
N_steps = 25000*2
N_walkers = 250*4
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

6.58 s ± 73.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [16]:
# Monte Carlo integration params
N_steps = 25000*2
N_walkers = 250*4
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

5.74 s ± 2.91 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [17]:
# Monte Carlo integration params
N_steps = 25000*3
N_walkers = 250*4
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

9.8 s ± 118 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [18]:
# Monte Carlo integration params
N_steps = 25000*3
N_walkers = 250*4
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

8.61 s ± 51.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Final remarks

Apart from the run time, another important aspect is the memory usage. Calculating more N_walkers and N_steps with more cores (for the same amount of time) also implies using more memory, which can be a limiting aspect. 

## Additional study after optimization

In [3]:
# Input parameters
E_local_f = E_local_Harmonic_Oscillator
prob_density = prob_density_Harmonic_Oscillator
dim = 1 # dimension of configuration space
alpha = np.array([0.65])
trial_move = 1.2222

N_skip = 1000
L_start = 5

In [4]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

2.34 s ± 248 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [5]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

2.22 s ± 180 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [10]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250
N_cores = 8

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

2.8 s ± 102 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [6]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*4
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

4.17 s ± 221 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [7]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*4
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

2.91 s ± 87.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [11]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*4
N_cores = 8

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

3.16 s ± 156 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [8]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*16
N_cores = 1

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

9.94 s ± 376 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [9]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*16
N_cores = 4

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

5.17 s ± 334 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [12]:
# Monte Carlo integration params
N_steps = 25000
N_walkers = 250*16
N_cores = 8

%timeit lib.MC_integration(E_local_f, prob_density, alpha, dim, N_steps=N_steps, N_walkers=N_walkers, N_skip=N_skip, L_start=L_start, N_cores=N_cores, trial_move=trial_move)

4.52 s ± 151 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Analysis of the N_walkers

Using N_steps = 25000, the results are

|BEFORE| N_walkers | time [s] (1 core) | time [s] (4 cores) |NOW | time [s] (1 core) | time [s] (4 cores) | time [s] (8 cores)
| --- | --- | --- |---|---| --- | --- | --- |
|| 250 | 1.75 | 2.07 |! | 2.34 | 2.22 |2.8|
|| 1000 | 3.76 | 3.02 |!| 4.12 | 2.91 |3.16|
|| 4000 | 9.75 | 6.38 |!| 9.94 | 5.17 |4.52|

The results are very similar, for high number of walkers the code is more efficient with 4 cores. The improvement gained has been of 19% in the best of cases, but we also lost efficiency in some others.